### PR TITLE
Use events instead of core code duplication

### PR DESCRIPTION
**Pull request by VSEphpbb**
This PR switches to core and template events, to inject the idea panel into the idea topic.

### CODE REVIEW

File: event/listener.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+				));
+			}
+		}
+
+		$points = $idea['idea_votes_up'] - $idea['idea_votes_down'];
+		$can_vote = (bool) ($idea['idea_status'] != ideas::STATUS_IMPLEMENTED &&
+			$idea['idea_status'] != ideas::STATUS_DUPLICATE &&
+			$this->auth->acl_get('f_vote', (int) $this->config['ideas_forum_id']) &&
+			$event['topic_data']['topic_status'] != ITEM_LOCKED);
+		$delete_posts = $mod || ($own && $this->auth->acl_get('f_delete', (int) $this->config['ideas_forum_id']));
```
Comment (michaelcullum) [inline event/listener.php]: Let's not let people delete their own idea, only m_

**Comment by michaelcullum** [comment]
Otherwise good to merge

**Comment by VSEphpbb** [comment]
Updated.

File: event/listener.php [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+		$idea = $this->ideas->get_idea_by_topic_id($event['topic_data']['topic_id']);
+
+		if (!$idea)
+		{
+			return;
+		}
+
+		$mod = $this->auth->acl_get('m_', (int) $this->config['ideas_forum_id']);
+		$own = $idea['idea_author'] === $this->user->data['user_id'];
```
Comment (rxu) [inline event/listener.php]: It might be a good idea to cast values to int here.

File: event/listener.php [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+			foreach ($statuses as $status)
+			{
+				$this->template->assign_block_vars('statuses', array(
+					'ID'	=> $status['status_id'],
+					'NAME'	=> $this->language->lang($status['status_name']),
+				));
+			}
+		}
+
+		$points = $idea['idea_votes_up'] - $idea['idea_votes_down'];
```
Comment (rxu) [inline event/listener.php]: Same as above to avoid strings substraction?

File: event/listener.php [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+			'S_CAN_VOTE'		=> $can_vote,
+
+			'U_DELETE_IDEA'		=> $this->link_helper->get_idea_link($idea['idea_id'], 'delete'),
+			'U_CHANGE_STATUS'	=> $this->link_helper->get_idea_link($idea['idea_id'], 'status', true),
+			'U_EDIT_DUPLICATE'	=> $this->link_helper->get_idea_link($idea['idea_id'], 'duplicate', true),
+			'U_EDIT_RFC'		=> $this->link_helper->get_idea_link($idea['idea_id'], 'rfc', true),
+			'U_EDIT_TICKET'		=> $this->link_helper->get_idea_link($idea['idea_id'], 'ticket', true),
+			'U_EDIT_TITLE'		=> $this->link_helper->get_idea_link($idea['idea_id'], 'title', true),
+			'U_REMOVE_VOTE'		=> $this->link_helper->get_idea_link($idea['idea_id'], 'removevote', true),
+			'U_IDEA_VOTE'		=> $this->link_helper->get_idea_link($idea['idea_id'], 'vote', true),
```
Comment (rxu) [inline event/listener.php]: We cast the ```get_idea_link()``` param to ```int``` above, probably shouldn't or cast it everywhere for consistency.

File: factory/ideas.php [hunk 5]
```diff
@@ -265,9 +253,29 @@ public function get_idea($id)
 	}
 
 	/**
+	 * Returns an idea specified by its topic ID.
+	 *
+	 * @param int $id The ID of the idea to return.
+	 *
+	 * @return array The idea
+	 */
+	public function get_idea_by_topic_id($id)
```
Comment (rxu) [inline factory/ideas.php]: We're adding 2 queries per every viewtopic page here, probably we could cache at least the following one. Alternatively, cache all ideas IDs and get the one needed from it if exists.

File: event/listener.php (hunk 3 above)
Comment (VSEphpbb) [inline event/listener.php]: They are already integers though (the DB is int)

File: styles/prosilver/template/event/viewtopic_body_post_buttons_list_before.html [hunk 6]
```diff
@@ -0,0 +1 @@
+<!-- DEFINE $SHOW_POST_BUTTONS = (postrow.POST_NUMBER != 1) -->
```
Comment (rxu) [inline styles/prosilver/template/event/viewtopic_body_post_buttons_list_before.html]: I guess it should be ```DEFINE $SHOW_POST_BUTTONS = ($SHOW_POST_BUTTONS and postrow.POST_NUMBER != 1)```, otherwise the post button will be shown with no respect of ```postrow.U_EDIT or postrow.U_DELETE or postrow.U_REPORT or postrow.U_WARN or postrow.U_INFO or postrow.U_QUOTE``` condition.

**Comment by rxu** [comment]
:+1:

File: factory/ideas.php (hunk 5 above)
Comment (VSEphpbb) [inline factory/ideas.php]: If these were not a query searching "where topic_id = id" then sure... But this query can have multiple variations and so its not really cacheable as each cache would be overwritten by the next idea viewed. It's just 1 extra query here, and its a tiny query searching for an index so it should be ok.

Comment (rxu) [inline factory/ideas.php]: 1 here plus another one from ```get_idea``` below it.

File: styles/prosilver/template/event/viewtopic_body_post_buttons_list_before.html (hunk 6 above)
Comment (VSEphpbb) [inline styles/prosilver/template/event/viewtopic_body_post_buttons_list_before.html]: I'm not sure what you mean? The idea is to not show any buttons in the first post, because the first post is the user bot.

We are inside a postrow loop when defining this, so we want to override it for the first post and force it set to false.

Comment (VSEphpbb) [inline styles/prosilver/template/event/viewtopic_body_post_buttons_list_before.html]: Oh think I see what you are saying...hmm..

File: factory/ideas.php (hunk 5 above)
Comment (VSEphpbb) [inline factory/ideas.php]: Right, it's just one tiny extra query... We eliminated 3 other queries though, so still a net savings going on ;)

An alternative may be to add topic_id param to the get_idea function, and search be either idea_id or topic_id but that can be a different PR if we really want it.

File: event/listener.php (hunk 4 above)
Comment (VSEphpbb) [inline event/listener.php]: it was not needed above so removed.

**Comment by VSEphpbb** [comment]
Updated

**Comment by VSEphpbb** [comment]
Updated
